src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,447 @@
+
+/--
+Traverse the expression 'e', and ackermannize potential subexpressions.
+We explicitly do not want to use `forallTelescoping` and family,
+because we want to only work with the zeroth order fragment (predicate logic),
+and ignore anything that is first (and higher) order.
+
+Hence, we use an `AckResult`, which keeps track of whether the resulting
+expression has a loose bvar.
+
+If the expression `e` is an application `f x₁ ... xₙ`, and none of `f`, `x₁`, ..., `xₙ`
+have loose bvars, we try to ackermannize at `ackApp`.
+If the application does have a bvar, we do not ackermannize.
+
+This invariant is maintained by `ackApp`, and `introAckForExpr` visits
+expressions and ackermannizes them in turn.
+-/
+
+partial def introAckForExpr (g : MVarId) (e : Expr) : AckM (AckResult × MVarId) :=
+  withContext g do
+    withTraceNode g (m!"🎯 {e}") (collapsed := false) do
+    match e with
+    | .mdata _ e => introAckForExpr g e
+    | .bvar _ => return ({ val := e, hasLooseBvar := true }, g)
+    | .proj tyName struct e =>
+      let (out, g) ← introAckForExpr g e
+      return ({ val := .proj tyName struct out.val, hasLooseBvar := out.hasLooseBvar}, g)
+    | .fvar .. | .mvar .. | .sort .. | .const .. |  .lit .. => do
+      return ({ val := e, hasLooseBvar := false }, g)
+    | .lam _binderName binderTy body _binderInfo =>
+        let (binderTy, g) ← introAckForExpr g binderTy
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateLambdaE! binderTy.val body.val
+        let hasLooseBvar := binderTy.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .letE _declName type value body _nonDep =>
+        let (type, g) ← introAckForExpr g type
+        let (value, g) ← introAckForExpr g value
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateLet! type.val value.val body.val
+        let hasLooseBvar := type.hasLooseBvar || value.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .forallE _binderName binderTy body _binderInfo =>
+        let (binderTy, g) ← introAckForExpr g binderTy
+        let (body, g) ← introAckForExpr g body
+        let val := e.updateForallE! binderTy.val body.val
+        let hasLooseBvar := binderTy.hasLooseBvar || body.hasLooseBvar
+        return ({ val, hasLooseBvar}, g)
+    | .app .. => withTraceNode g m!"🎯 Expr.app '{e}'" (collapsed := false) do
+      ackApp g e
+end
+
+/--
+Return true if the argument lists are trivially different.
+This is an optimization that we do not yet implement.
+-/
+def areArgListsTriviallyDifferent (_arg₁ _arg₂ : Array Argument) : AckM Bool := return false
+
+/-
+Return true if the argument lists are trivially the same.
+This is an optimization that we do not yet implement.
+If possible, return the simplified hypothesis of the equality of these expressions.
+TODO: -- def areArgListsTriviallySame (arg₁ arg₂ : Array Argument) : AckM (Option Expr) := return none
+-/
```

Comment:
```suggestion
```

Do you have plans to implement this optimization soon? If not, adding these stubs just feels a bit too noisy for me, and I'd prefer a comment at the use-site explaining that we can bail out early if the arguments lists are different.